# Recommendation System

A recommender system or a recommendation system (sometimes replacing "system" with a synonym such as platform or engine) is a subclass of information filtering system that seeks to predict the "rating" or "preference" a user would give to an item.They are primarily used in commercial applications.

Recommender systems are utilized in a variety of areas, and are most commonly recognized as playlist generators for video and music services like Netflix, YouTube and Spotify, product recommenders for services such as Amazon, or content recommenders for social media platforms such as Facebook and Twitter. These systems can operate using a single input, like music, or multiple inputs within and across platforms like news, books, and search queries. There are also popular recommender systems for specific topics like restaurants and online dating. Recommender systems have been developed to explore research articles and experts, collaborators, financial services and life insurance.

In [3]:
import pandas as pd
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import re
import html
from html.parser import HTMLParser
import os
import warnings
warnings.filterwarnings('ignore')

In [6]:
os.chdir(r'C:\Users\ganes\Downloads\ML JOB\Coding')

In [7]:
artists = pd.read_csv('artist.csv', encoding='latin-1')
artists = artists.rename(columns={'id': 'artistID'})
artists.head()

artistID               name
0         1       MALICE MIZER
1         2    Diary of Dreams
2         3  Carpathian Forest
3         4       Moi dix Mois
4         5        Bella Morte

In [8]:
artists.shape

(17632, 2)

In [9]:
# Cleaning artist name
def clean_column(column):
    hparser=HTMLParser()  
    new_text=html.unescape(column)  # converts HTML character codes to ASCII equivalent
    lower_case = new_text.lower()  # converts all the letters to lower case
    spaces = re.sub('[\s]+', ' ', lower_case) # Removes the spaces
    link = re.sub('((www\.[^\s]+)|(https?://[^\s]+))','',spaces) # Removes the URL's if it finds the lines with http. or www.
    alpha_numeric = re.sub('[^a-zA-Z\s]+', '', link) # Removes the alpha numeric symbols
    return alpha_numeric

clean_artist_name = artists['name'].apply(lambda column: clean_column(column)) # Applies above changes made to all the rows in the data set
artists['cleaned_artist_name'] = clean_artist_name

# Replacing empty strings with Nan
artists['cleaned_artist_name'].replace("", np.nan, inplace=True)
artists['cleaned_artist_name'].replace(" ", np.nan, inplace=True)
artists['cleaned_artist_name'].replace("  ", np.nan, inplace=True)
artists['cleaned_artist_name'].replace("   ", np.nan, inplace=True)
artists['cleaned_artist_name'].replace("    ", np.nan, inplace=True)

In [10]:
artists['cleaned_artist_name'].head(10)

0         malice mizer
1      diary of dreams
2    carpathian forest
3         moi dix mois
4          bella morte
5            moonspell
6       marilyn manson
7          dir en grey
8          combichrist
9              grendel
Name: cleaned_artist_name, dtype: object

In [11]:
artists['cleaned_artist_name'].shape

(17632,)

In [12]:
# artists['cleaned_artist_name'].isnull().sum(axis=0)

In [46]:
user_artist = pd.read_csv('user_artist.csv')
user_artist

userID  artistID  weight
0           2        51   13883
1           2        52   11690
2           2        53   11351
3           2        54   10300
4           2        55    8983
5           2        56    6152
6           2        57    5955
7           2        58    4616
8           2        59    4337
9           2        60    4147
10          2        61    3923
11          2        62    3782
12          2        63    3735
13          2        64    3644
14          2        65    3579
15          2        66    3312
16          2        67    3301
17          2        68    2927
18          2        69    2720
19          2        70    2686
20          2        71    2654
21          2        72    2619
22          2        73    2584
23          2        74    2547
24          2        75    2397
25          2        76    2382
26          2        77    2120
27          2        78    2119
28          2        79    1990
29          2        80    1972
30          2        81    1948
31          2        82    1868
32          2        83    1792
33          2        84    1740
34          2        85    1638
35          2        86    1594
36          2        87    1559
37          2        88    1553
38          2        89    1519
39          2        90    1471
40          2        91    1438
41          2        92    1411
42          2        93    1407
43          2        94    1373
44          2        95    1363
45          2        96    1342
46          2        97    1337
47          2        98    1332
48          2        99    1330
49          2       100    1315
50          3       101   13176
51          3       102     662
52          3       103     493
53          3       104     431
54          3       105     403
55          3       106     354
56          3       107     269
57          3       108     236
58          3       109     215
59          3       110     215
60          3       111     212
61          3       112     193
62          3       113     187
63          3       114     167
64          3       115     163
65          3       116     154
66          3       117     151
67          3       118     142
68          3       119     142
69          3       120     125
70          3       121     111
71          3       122     108
72          3       123     104
73          3       124      99
74          3       125      98
75          3       126      94
76          3       127      89
77          3       128      89
78          3       129      86
79          3       130      85
80          3       131      84
81          3       132      83
82          3       133      83
83          3       134      77
84          3       135      77
85          3       136      76
86          3       137      75
87          3       138      72
88          3       139      72
89          3       140      71
90          3       141      70
91          3       142      70
92          3       143      70
93          3       144      69
94          3       145      68
95          3       146      67
96          3       147      67
97          3       148      66
98          3       149      66
99          3       150      65
100         4        51     228
101         4        53     686
102         4        64     420
103         4        65     301
104         4        70     389
105         4        72    4983
106         4        77     396
107         4       151    1807
108         4       152    1208
109         4       153     903
110         4       154     826
111         4       155     777
112         4       156     743
113         4       157     732
114         4       158     729
115         4       159     724
116         4       160     712
117         4       161     632
118         4       162     524
119         4       163     465
120         4       164     455
121         4       165     446
122         4       166     445
123         4       167     407
124    

In [14]:
# Merging two dataframes
df_ratings = user_artist.join(artists.set_index('artistID'), on='artistID')
df_ratings = df_ratings.rename(columns={'name': 'artist_name'})
df_ratings.head()

userID  artistID  weight    artist_name cleaned_artist_name
0       2        51   13883    Duran Duran         duran duran
1       2        52   11690      Morcheeba           morcheeba
2       2        53   11351            Air                 air
3       2        54   10300   Hooverphonic        hooverphonic
4       2        55    8983  Kylie Minogue       kylie minogue

In [15]:
drop_columns = ['weight', 'artist_name']
df_ratings = df_ratings.drop(drop_columns, axis = 1)
df_ratings.head()

userID  artistID cleaned_artist_name
0       2        51         duran duran
1       2        52           morcheeba
2       2        53                 air
3       2        54        hooverphonic
4       2        55       kylie minogue

In [16]:
df_ratings['cleaned_artist_name'].value_counts()

lady gaga                                   611
britney spears                              525
rihanna                                     485
the beatles                                 480
katy perry                                  474
madonna                                     429
avril lavigne                               417
christina aguilera                          407
muse                                        400
paramore                                    400
beyonc                                      397
radiohead                                   393
coldplay                                    369
keha                                        362
shakira                                     319
pnk                                         305
the killers                                 304
black eyed peas                             304
kylie minogue                               298
miley cyrus                                 287
depeche mode                            

I would like to rate the artists based on how many times they have rated by the users. It seems like Lady Gaga has been rated most no of times which is 611 and there are many other artists who were rated more than 200 times. For artists who were rated more than 400 times consider them as 5 star artists and artists with 300 and more #ratings consider them as 4 star artists and artists with 200 and more #ratings consider them as 3 star artists and artists with 100 and more #ratings are considered as 2 star artists and artists with less than 100 #ratings are considered as 1 star artists.

In [17]:
pd.set_option('display.max_rows', 500)
df_ratings['# artist_rated'] = df_ratings.groupby(['artistID'])['cleaned_artist_name'].transform('count')
df_ratings.head()

userID  artistID cleaned_artist_name  # artist_rated
0       2        51         duran duran             111
1       2        52           morcheeba              23
2       2        53                 air              75
3       2        54        hooverphonic              18
4       2        55       kylie minogue             298

Now let's rate the artists based on the count of each artist

In [18]:
df_ratings['rating'] = df_ratings['# artist_rated'].apply(lambda x:5 if x >= 400 else 4 if x >= 300 else 3 if x >= 200 
                                                          else 2 if x >= 100 else 1) 

In [19]:
df_ratings.head()

userID  artistID cleaned_artist_name  # artist_rated  rating
0       2        51         duran duran             111       2
1       2        52           morcheeba              23       1
2       2        53                 air              75       1
3       2        54        hooverphonic              18       1
4       2        55       kylie minogue             298       3

In [20]:
df_ratings['rating'].value_counts()

1    67925
2    11224
3     6310
5     4223
4     3152
Name: rating, dtype: int64

### Building a simple recommendation system

One approach to the design of recommender systems that has wide use is collaborative filtering. Collaborative filtering is based on the assumption that people who agreed in the past will agree in the future, and that they will like similar kinds of items as they liked in the past. The system generates recommendations using only information about rating profiles for different users or items.

### User-User Collaborative filtering

This algorithm first finds the similarity score between users. Based on this similarity score, it then picks out the most similar users and recommends products which these similar users have liked or bought previously.

![image](https://s3-ap-south-1.amazonaws.com/av-blog-media/wp-content/uploads/2018/05/0o0zVW2O6Rv-LI5Mu1-850x466.png)

### Item-Item Collaborative filtering

In this algorithm, we compute the similarity between each pair of items

![image](https://s3-ap-south-1.amazonaws.com/av-blog-media/wp-content/uploads/2018/05/1skK2fqWiBF7weHU8SjuCzw.png)

In [21]:
n_users = df_ratings.userID.unique().shape[0]
n_items = df_ratings.artistID.unique().shape[0]

In [22]:
data_matrix = np.zeros((df_ratings['userID'].max(), df_ratings['artistID'].max()))

#populate the matrix based on the dataset
for line in df_ratings.itertuples():
    data_matrix[line[1]-1, line[2]-1] = line[5]


In [23]:
# data_matrix.unique.shape[0]
np.unique(data_matrix)

array([0., 1., 2., 3., 4., 5.])

In [24]:
data_matrix_df = pd.DataFrame(data_matrix)
data_matrix_df.head()

0      1      2      3      4      5      6      7      8      9      \
0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

   ...    18735  18736  18737  18738  18739  18740  18741  18742  18743  18744  
0  ...      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
1  ...      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
2  ...      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
3  ...      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
4  ...      0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 18745 columns]

Now, we will calculate the similarity. We can use the pairwise_distance function from sklearn to calculate the cosine similarity. This gives us the item-item and user-user similarity in an array form. The next step is to make predictions based on these similarities. Let’s define a function to do just that.

In [25]:
from sklearn.metrics.pairwise import pairwise_distances 
user_similarity = pairwise_distances(data_matrix, metric='cosine')
item_similarity = pairwise_distances(data_matrix.T, metric='cosine')

In [26]:
user_similarity_dataframe = pd.DataFrame(user_similarity)
user_similarity_dataframe.head()

0         1     2         3         4         5         6         7     \
0   0.0  1.000000   1.0  1.000000  1.000000  1.000000  1.000000  1.000000   
1   1.0  0.000000   1.0  0.777676  0.994236  0.989058  0.732963  0.855662   
2   1.0  1.000000   0.0  1.000000  1.000000  1.000000  1.000000  1.000000   
3   1.0  0.777676   1.0  0.000000  0.692430  0.989385  1.000000  1.000000   
4   1.0  0.994236   1.0  0.692430  0.000000  1.000000  1.000000  1.000000   

   8         9     ...       2090      2091      2092      2093  2094  \
0   1.0  1.000000  ...   1.000000  1.000000  1.000000  1.000000   1.0   
1   1.0  1.000000  ...   0.913207  1.000000  0.990247  0.843848   1.0   
2   1.0  1.000000  ...   1.000000  1.000000  1.000000  1.000000   1.0   
3   1.0  0.817761  ...   1.000000  0.936908  0.990538  1.000000   1.0   
4   1.0  0.644492  ...   1.000000  0.538160  1.000000  1.000000   1.0   

       2095      2096  2097  2098  2099  
0  1.000000  1.000000   1.0  1.00   1.0  
1  0.972507  0.862081   1.0  1.00   1.0  
2  1.000000  1.000000   1.0  0.98   1.0  
3  0.586584  0.719654   1.0  1.00   1.0  
4  0.549292  0.667460   1.0  1.00   1.0  

[5 rows x 2100 columns]

In [27]:
item_similarity_dataframe =  pd.DataFrame(item_similarity)
item_similarity_dataframe.head()

0         1      2      3         4         5         6         7      \
0    0.0  1.000000    1.0    1.0  1.000000  1.000000  0.949937  0.833333   
1    1.0  0.000000    1.0    1.0  0.795876  1.000000  0.924906  1.000000   
2    1.0  1.000000    0.0    1.0  1.000000  0.452277  1.000000  1.000000   
3    1.0  1.000000    1.0    0.0  1.000000  1.000000  0.938686  0.795876   
4    1.0  0.795876    1.0    1.0  0.000000  1.000000  0.938686  1.000000   

      8         9      ...    18735  18736  18737  18738  18739  18740  18741  \
0  1.000000  1.000000  ...      1.0    1.0    1.0    1.0    1.0    1.0    1.0   
1  0.646447  0.711325  ...      1.0    1.0    1.0    1.0    1.0    1.0    1.0   
2  1.000000  1.000000  ...      1.0    1.0    1.0    1.0    1.0    1.0    1.0   
3  1.000000  1.000000  ...      1.0    1.0    1.0    1.0    1.0    1.0    1.0   
4  0.855662  1.000000  ...      1.0    1.0    1.0    1.0    1.0    1.0    1.0   

   18742  18743  18744  
0    1.0    1.0    1.0  
1    1.0    1.0    1.0  
2    1.0    1.0    1.0  
3    1.0    1.0    1.0  
4    1.0    1.0    1.0  

[5 rows x 18745 columns]

In [28]:
def predict(data_matrix, similarity, type='user'):
    if type == 'user':
        mean_user_rating = data_matrix.mean(axis=1)
        #We use np.newaxis so that mean_user_rating has same format as ratings
        ratings_diff = (data_matrix - mean_user_rating[:, np.newaxis])
        ratings_diff = (data_matrix)
        pred = mean_user_rating[:, np.newaxis] + similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
#         pred = similarity.dot(ratings_diff) / np.array([np.abs(similarity).sum(axis=1)]).T
    elif type == 'item':
        pred = data_matrix.dot(similarity) / np.array([np.abs(similarity).sum(axis=1)])
    return pred

In [29]:
user_prediction = predict(data_matrix, user_similarity, type='user')
item_prediction = predict(data_matrix, item_similarity, type='item')

In [30]:
user_prediction.shape

(2100, 18745)

In [31]:
item_prediction.shape

(2100, 18745)

In [32]:
user_prediction_df = pd.DataFrame(user_prediction)
user_prediction_df.head()

0         1         2         3         4         5         6      \
0  0.001429  0.005717  0.001429  0.000953  0.000953  0.004764  0.126727   
1  0.005235  0.009589  0.005338  0.004766  0.004761  0.008970  0.128870   
2  0.004097  0.008387  0.004097  0.003621  0.003621  0.007434  0.129455   
3  0.005197  0.009375  0.005494  0.004839  0.004705  0.008800  0.124106   
4  0.005902  0.010415  0.006293  0.005601  0.005499  0.009725  0.120328   

      7         8         9        ...        18735     18736     18737  \
0  0.005717  0.011434  0.004288    ...     0.000476  0.000476  0.000476   
1  0.009592  0.015298  0.008339    ...     0.004269  0.004269  0.004269   
2  0.008387  0.014107  0.006957    ...     0.003144  0.003144  0.003144   
3  0.009524  0.014496  0.008276    ...     0.004429  0.004429  0.004429   
4  0.010401  0.015905  0.009025    ...     0.005171  0.005171  0.005171   

      18738     18739     18740     18741     18742     18743     18744  
0  0.000476  0.000476  0.000476  0.000476  0.000476  0.000476  0.000476  
1  0.004269  0.004269  0.004269  0.004269  0.004269  0.004269  0.004208  
2  0.003144  0.003144  0.003144  0.003144  0.003144  0.003144  0.003144  
3  0.004429  0.004429  0.004429  0.004429  0.004429  0.004429  0.004262  
4  0.005171  0.005171  0.005171  0.005171  0.005171  0.005171  0.004940  

[5 rows x 18745 columns]

In [33]:
item_prediction_df = pd.DataFrame(item_prediction)
item_prediction_df.head()

0         1         2         3         4         5         6      \
0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
1  0.003716  0.003698  0.003743  0.003734  0.003732  0.003738  0.003555   
2  0.002672  0.002677  0.002674  0.002672  0.002671  0.002676  0.002689   
3  0.003847  0.003819  0.003900  0.003856  0.003829  0.003857  0.003615   
4  0.004570  0.004593  0.004653  0.004615  0.004598  0.004624  0.004201   

      7         8         9        ...        18735     18736     18737  \
0  0.000000  0.000000  0.000000    ...     0.000000  0.000000  0.000000   
1  0.003710  0.003665  0.003732    ...     0.003740  0.003740  0.003740   
2  0.002676  0.002681  0.002677    ...     0.002671  0.002671  0.002671   
3  0.003834  0.003754  0.003848    ...     0.003900  0.003900  0.003900   
4  0.004564  0.004540  0.004601    ...     0.004638  0.004638  0.004638   

      18738     18739     18740     18741     18742     18743     18744  
0  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  
1  0.003740  0.003740  0.003740  0.003740  0.003740  0.003740  0.003708  
2  0.002671  0.002671  0.002671  0.002671  0.002671  0.002671  0.002670  
3  0.003900  0.003900  0.003900  0.003900  0.003900  0.003900  0.003856  
4  0.004638  0.004638  0.004638  0.004638  0.004638  0.004638  0.004570  

[5 rows x 18745 columns]

### Getting recommendations based on item and user similarity

In [49]:
#Construct a reverse map of indices and artist titles

index=df_ratings['cleaned_artist_name'].drop_duplicates()
index
indices = pd.Series(index)
indices = indices.tolist()

In [50]:
# Function that takes in artist title as input and outputs most similar artist
def get_recommendations(cleaned_artist_name, cosine_sim= item_similarity):
    # Get the index of the artist that matches the title
    idx = indices.index(cleaned_artist_name)

    # Get the pairwsie similarity scores of all artists with that artist
    sim_scores = list(enumerate(item_similarity[idx]))

    # Sort the artists based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar artists
    sim_scores = sim_scores[1:11]

    # Get the artist indices
    artist_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar artists
    return df_ratings['cleaned_artist_name'].iloc[artist_indices]

In [53]:
get_recommendations('coldplay')

1         morcheeba
3      hooverphonic
4     kylie minogue
7         goldfrapp
13          ryksopp
15        faithless
18             sade
19             moby
20             dido
21     depeche mode
Name: cleaned_artist_name, dtype: object

In [55]:
user_index=df_ratings['userID'].drop_duplicates()
user_index
indices_user = pd.Series(user_index)
indices_user = indices_user.tolist()

In [56]:
# Function that takes in userID as input and outputs most similar movies
def get_recommendations(userID, cosine_sim= user_similarity):
    # Get the index of the userID that matches the title
    idx = indices_user.index(userID)

    # Get the pairwsie similarity scores of all artists with that artist
    sim_scores = list(enumerate(user_similarity[idx]))

    # Sort the userID based on the similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar users
    sim_scores = sim_scores[1:11]

    # Get the user indices
    user_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df_ratings['cleaned_artist_name'].iloc[user_indices]

In [57]:
get_recommendations(2100)

2                air
5          daft punk
15         faithless
18              sade
19              moby
20              dido
21      depeche mode
27    the adventures
32        cock robin
34    spandau ballet
Name: cleaned_artist_name, dtype: object